Configure Blaze: enable BLAS kernels, col-major matrix storage, add comments

nilsvu opened the pull request:
## Proposed changes

Update the Blaze configuration. In particular, I found odd that external BLAS kernels were not enabled.

- Enable external BLAS kernels
- Set default matrix storage order to column-major
- ~Enable padding (Blaze default). It's recommended by Blaze for performance, and is already supported by most of our functions since Blaze 3.7 had padding always enabled.~ moved to #3825 
- Remove some redundant configuration flags

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu commented on src/Evolution/DgSubcell/Matrices.cpp:
```diff
@@ -283,9 +283,10 @@ Matrix reconstruction_matrix_cache_impl_helper(
   // than Blaze.
```
>   // Do matrix multipy with dgemm_ directly because that seems to be faster
>   // than Blaze

@nilsdeppe The reason for this could be that we hadn't enabled BLAS kernels for Blaze. You could check again now that it's enabled.

nilsdeppe replied:
Isn't Blaze hard-coded to use cblas? I'm not sure how the header choice effects the implementation, but cblas is pretty much always the slowest blas implementation possible. In general we'll want to use MKL or OpenBLAS.

nilsvu replied:
Blaze uses whichever BLAS you link in. The header doesn't affect this, it only defines the interface (standardized by BLAS). That's why many people don't even bother including the header but call the BLAS functions "blindly". The Blaze docs say you should enable BLAS for max performance; in particular the BLAS dense matrix multiplication kernel is faster than Blaze's fallback: https://bitbucket.org/blaze-lib/blaze/wiki/Configuration%20and%20Installation#!step-2-configuration

That could be the reason why you found Blaze performed worse than BLAS for this particular case. With BLAS enabled in Blaze you should be able to use matrix multiplication directly here.

Note also that "CBLAS" is not a particular BLAS implementation, it only refers to the C bindings to the BLAS functions, which are provided by OpenBLAS, MKL, etc. This is my understanding.

wthrowe commented:
Tests are failing.

nilsvu commented:
I must have missed one in the subcell code. Will find it tomorrow.

nilsvu commented on src/Evolution/DgSubcell/Matrices.cpp:
```diff
@@ -229,13 +229,13 @@ Matrix reconstruction_matrix_cache_impl_helper(
   for (size_t l = 0; l < num_pts; ++l) {
     for (size_t j = 0; j < num_pts; ++j) {
-      lhs_recons_matrix(l, j) = *(rhs_recons_matrix.data() + (j + l * num_pts));
+      lhs_recons_matrix(l, j) = rhs_recons_matrix(j, l);
```
Alternatives, if preferred:

- Add spacing to the pointer arithmetic
- Use `blaze::submatrix` and `blaze::trans` instead of the loops

See also https://github.com/sxs-collaboration/spectre/pull/3806#discussion_r807778323

nilsdeppe replied:
This should be fine, I think

nilsdeppe commented on cmake/SetupBlaze.cmake:
```diff
@@ -19,24 +19,32 @@ set_property(TARGET Blaze PROPERTY
+  # Enable external BLAS kernels
+  BLAZE_BLAS_MODE=1
+  # Set default matrix storage order to column-major
+  BLAZE_DEFAULT_STORAGE_ORDER=blaze::columnMajor
```
This changes the storage order. Why?

nilsvu replied:
Things like partial derivatives, `apply_matrices` etc are implemented for col-major layout. With this config you get col-major matrices from Blaze by default, making it easier to work with those functions. In fact, currently we have code like this in the repo:

```cpp
  // src/Evolution/Systems/NewtonianEuler/Characteristics.cpp
  const Matrix a_x = blaze::DynamicMatrix<double>{
      {0., 1., 0.},
      {b_times_theta - square(u), u * (2. - kappa_over_density),
       kappa_over_density}, ...};
```

Here `Matrix` is col-major and `blaze::DynamicMatrix` is row-major (current default). The storage order could be made explicit here, but switching the default to col-major seems sensible to me to make things consistent. I think we can get rid of the shallow `Matrix` wrapper class very easily with the col-major default.

I split this change into a separate commit.

nilsdeppe replied:
okay, that sounds reasonable. Thanks!

nilsdeppe commented on cmake/SetupBlaze.cmake:
```diff
@@ -19,24 +19,32 @@ set_property(TARGET Blaze PROPERTY
+  BLAZE_DEFAULT_STORAGE_ORDER=blaze::columnMajor
+  # Disable SMP parallelization
   BLAZE_USE_SHARED_MEMORY_PARALLELIZATION=0
-  BLAZE_OPENMP_PARALLEL_MODE=0
```
Please keep all parallelization things explicitly disabled. We've had issues with blas libraries doing multithreading on us in the past.

nilsvu replied:
The `BLAZE_USE_SHARED_MEMORY_PARALLELIZATION` switch disables SMP parallelization altogether, see: https://bitbucket.org/blaze-lib/blaze/wiki/Serial%20Execution#!option-3-deactivation-of-parallel-execution

I split this change into a separate commit.

nilsdeppe commented on cmake/SetupBlaze.cmake:
```diff
@@ -19,24 +19,32 @@ set_property(TARGET Blaze PROPERTY
   BLAZE_MPI_PARALLEL_MODE=0
-  # Disable HPX parallelization
-  BLAZE_HPX_PARALLEL_MODE=0
-  BLAZE_USE_PADDING=0
```
let's not enable padding... That is likely going to cause headaches and there's no data indicating it'll improve anything.

nilsvu replied:
Blaze warns that disabling padding can considerably decrease performance for dense matrix operations: https://bitbucket.org/blaze-lib/blaze/src/c4d9e85414370e880e5e79c86e3c8d4d38dcde7a/blaze/config/Optimizations.h#lines-52

To support padding, all you have to do is use `.spacing()` instead of `.rows()/.columns()` in the `LDA`, `LDB`, etc. parameters of LAPACK.

I split this change into a separate commit.

nilsdeppe replied:
I don't know how this translates to LIBXSMM. Naively I assume it would be similar speedups, but I think it would be good to do a microbenchmark on a machine with AVX or AVX2 support since that's where I expect the impact will be most noticeable. The reason being that unaligned accesses used to be very slow with vectorization but CPUs over the last 4-5 years don't have that issue anymore, but I don't know what the case is with padding. Does that sound okay?

nilsdeppe commented on cmake/SetupBlaze.cmake:
```diff
@@ -19,24 +19,32 @@ set_property(TARGET Blaze PROPERTY
-  BLAZE_HPX_PARALLEL_MODE=0
-  BLAZE_USE_PADDING=0
-  # Enable non-temporal stores for cache optimization of large data structures
-  BLAZE_USE_STREAMING=1
```
please keep these optimizations enabled

nilsvu replied:
These are enabled by default. I kept `BLAZE_USE_STREAMING` explicitly enabled for you. I think it's not really necessary, but I don't feel strongly about it. I split this change into a separate commit.

nilsdeppe commented on cmake/SetupGsl.cmake:
```diff
@@ -12,6 +12,13 @@ file(APPEND
   "GSL version: ${GSL_VERSION}\n"
   )
 
+# Link external BLAS library. We don't need the GSL::gslcblas target.
```
This change makes no sense to me. Why would GSL depend on Blas?

nilsvu replied:
GSL depends on BLAS 🤷‍♂️ https://www.gnu.org/software/gsl/doc/html/usage.html#linking-programs-with-the-library:

> To link against the library you need to specify both the main library and a supporting CBLAS library, which provides standard basic linear algebra subroutines. A suitable CBLAS implementation is provided in the library libgslcblas.a if your system does not provide one.

Note that with "CBLAS" they mean the C bindings to the BLAS functions, as provided by any BLAS implementation like OpenBLAS or MKL (see also https://github.com/sxs-collaboration/spectre/pull/3806#discussion_r808804170).

nilsdeppe replied:
ah, okay, that makes sense. Thanks!

nilsdeppe commented on cmake/SetupBlaze.cmake:
```diff
@@ -19,24 +19,32 @@ set_property(TARGET Blaze PROPERTY
+  # implementation, but it's not guaranteed to be available and may conflict
+  # with the GSL header. Since we use GSL anyway, it's easier to use their BLAS
+  # header.
+  BLAZE_BLAS_INCLUDE_FILE=<gsl/gsl_cblas.h>
```
what are the performance implications of using the cblas header? Does this mean Blaze will use CBlas instead of MKL or OpenBLAS?

nilsvu replied:
See comment above: https://github.com/sxs-collaboration/spectre/pull/3806#discussion_r808804170

nilsvu committed:
Blaze: enable BLAS kernels

Blaze uses BLAS kernels for max performance; in particular the
BLAS dense matrix multiplication kernel is faster than Blaze's fallback
according to their docs:
https://bitbucket.org/blaze-lib/blaze/wiki/Configuration%20and%20Installation#!step-2-configuration
Therefore, we follow the Blaze recommendation and make sure BLAS
is enabled and linked.

Also make sure LAPACK is linked, which is a dependency of Blaze.

nilsvu committed:
gsl: link BLAS dependency

nilsvu committed:
Blaze: set default matrix storage order to column-major

Many of our functions are implemented for column-major layout. Setting
this default reduces conversion between the data layouts.

nilsvu committed:
Blaze: disable SMP parallelization

The `BLAZE_USE_SHARED_MEMORY_PARALLELIZATION` disables
SMP parallelization for all possible backends
(OpenMP, C++11 threads, Boost, HPX), so the individual flags are
not needed:
https://bitbucket.org/blaze-lib/blaze/wiki/Serial%20Execution#!option-3-deactivation-of-parallel-execution

nilsvu committed:
Blaze: add notes on machine-specific options

nilsvu committed:
Blaze: add notes on optimization flags

Also remove `BLAZE_USE_OPTIMIZED_KERNELS`, which is enabled by default.
`BLAZE_USE_STREAMING` is also enabled by default, but we keep it
explicitly listed to make sure it's always enabled.

nilsvu commented:
I split this PR into a series of commits so we can more easily reason about each configuration. I hope this clarifies some of the config choices.

nilsdeppe commented on cmake/SetupBlaze.cmake:
```diff
@@ -55,10 +55,12 @@ target_compile_definitions(Blaze
   BLAZE_USE_STREAMING=1
-  BLAZE_USE_OPTIMIZED_KERNELS=1
-  # Skip initializing default-constructed structures for fundamental types
+  # - We could use SLEEF for vectorized implementations of sin, cos, exp, etc.:
```
Just an FYI that I was looking at this and have a branch somewhere that adds sleef support. I'll submit that once I get some time :)

nilsvu replied:
nice 👍

nilsdeppe reviewed (commented):
Just the one timing measurement request. I am a bit worried that padding might cause some weird FPEs on certain machines (I hope not). We saw this with SpEC on Stampede2 and I think Frontera, so it might be worth testing this on one or both of those machines just to be sure :)

nilsvu commented:
@wthrowe @nilsdeppe Running the `Benchmark` executable on develop on Frontera gives these errors:

```
login4.frontera(1066)$ ./bin/Benchmark 
[1645185543.861533] [login4:196244:0]        rc_iface.c:506  UCX  WARN  using maximal value for RETRY_COUNT (7) instead of 40
[1645185543.864173] [login4:196244:0]        rc_iface.c:506  UCX  WARN  using maximal value for RETRY_COUNT (7) instead of 40
------- Partition 0 Processor 0 Exiting: Called CmiAbort ------
Reason: UCX: UcxInitEps: runtime_kvs_put error failed: 5
[0] Stack Traceback:
  [0:0] Benchmark 0x62ce4b 
  [0:1] Benchmark 0x62cf60 
  [0:2] Benchmark 0x63164c 
  [0:3] Benchmark 0x632f86 
  [0:4] Benchmark 0x534f47 
  [0:5] libc.so.6 0x2abbd9bf0555 __libc_start_main
  [0:6] Benchmark 0x43391e 
```

I see `UCX` output, so is charm stuff interfering here for some reason? @wthrowe I remember you had to run tests with `charmrun` on Frontera, but the `Benchmark` executable should have nothing to do with charm so I'm not sure this is related.. any ideas?

wthrowe commented:
Did you try running it through charmrun?

nilsvu commented:
@wthrowe I did. The Charm++ startup info was printed but then the program hangs.

nilsdeppe commented:
Should we leave the padding change for a followup PR (happy to be opened immediately!) and get everything else in? That way it'll minimize any possible conflicts and we can keep the PR moving quickly :)

nilsvu commented:
Sounds good! I removed the padding commits from this PR and opened #3825
